File: cpp/trees/binary_tree_traversal.cpp
/*
Problem: Binary Tree Inorder Traversal
Approaches:
1. Recursive
2. Iterative stack
3. Morris
*/
#include <bits/stdc++.h>
using namespace std;

struct TreeNode { int val; TreeNode* left; TreeNode* right; TreeNode(int v): val(v), left(nullptr), right(nullptr) {} };

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        // TODO: implement recursive then iterative
        return {}; // placeholder
    }
};

int main() {
    TreeNode* root = new TreeNode(1);
    root->right = new TreeNode(2);
    root->right->left = new TreeNode(3);
    Solution s; auto ans = s.inorderTraversal(root);
    for (int v: ans) cout << v << ' '; cout << endl; // expect 1 3 2
}

File: python/trees/binary_tree_traversal.py
"""
Problem: Binary Tree Inorder Traversal
Given the root of a binary tree, return the inorder traversal of its nodes' values.

Approaches:
1. Recursive DFS
2. Iterative stack
3. Morris traversal (no extra space)
"""
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        # TODO: recursive
        # TODO: iterative
        raise NotImplementedError

if __name__ == "__main__":
    root = TreeNode(1, None, TreeNode(2, TreeNode(3)))
    print(Solution().inorderTraversal(root))  # expect [1,3,2]
